lib/grant-types/authorization-code-grant-type.js

```diff
@@ -0,0 +1,170 @@
+
+/**
+ * Module dependencies.
+ */
+
+var AbstractGrantType = require('./abstract-grant-type');
+var InvalidArgumentError = require('../errors/invalid-argument-error');
+var InvalidGrantError = require('../errors/invalid-grant-error');
+var InvalidRequestError = require('../errors/invalid-request-error');
+var Promise = require('bluebird');
+var ServerError = require('../errors/server-error');
+var is = require('../validator/is');
+var util = require('util');
+
+/**
+ * Constructor.
+ */
+
+function AuthorizationCodeGrantType(options) {
+  options = options || {};
+
+  if (!options.model) {
+    throw new InvalidArgumentError('Missing parameter: `model`');
+  }
+
+  if (!options.model.getAuthorizationCode) {
+    throw new InvalidArgumentError('Invalid argument: model does not implement `getAuthorizationCode()`');
+  }
+
+  if (!options.model.revokeAuthorizationCode) {
```

Comment:
This is enforced by the OAuth RFC 6479, [section 4.1.2](https://tools.ietf.org/html/rfc6749#section-4.1.2):

> Authorization codes MUST be short lived and single-use. If an authorization code is used more than once, the authorization server MUST deny the request and SHOULD revoke (when possible) all tokens previously issued based on that authorization code.

Note that authorization codes _MUST_ be single-use, which is why revoking the code is necessary.